File: specter/tests/unit/parser/parse.test.ts
/**
 * Tests for spec-parse: YAML-to-SpecAST parser.
 *
 * @spec spec-parse
 *
 * Every test maps to an acceptance criterion in specs/spec-parse.spec.yaml.
 * See references_constraints on each AC for which constraints are validated.
 */

import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { parseSpec } from '../../../src/core/parser/parse.js';

const fixturesDir = resolve(import.meta.dirname, '../../fixtures');

function readFixture(relativePath: string): string {
  return readFileSync(resolve(fixturesDir, relativePath), 'utf-8');
}

describe('spec-parse', () => {
  // @ac AC-01: Valid spec file is parsed into a SpecAST with all required fields populated
  // References: C-01 (schema validation), C-04 (typed SpecAST output)
  it('AC-01: parses a valid spec into a SpecAST with all required fields', () => {
    const yaml = readFixture('valid/simple.spec.yaml');
    const result = parseSpec(yaml);

    expect(result.ok).toBe(true);
    if (!result.ok) return;

    const ast = result.value;
    expect(ast.id).toBe('test-simple');
    expect(ast.version).toBe('1.0.0');
    expect(ast.status).toBe('approved');
    expect(ast.tier).toBe(2);
    expect(ast.context).toBeDefined();
    expect(ast.context.system).toBe('Test system');
    expect(ast.objective).toBeDefined();
    expect(ast.objective.summary).toBeDefined();
    expect(ast.constraints).toHaveLength(1);
    expect(ast.constraints[0].id).toBe('C-01');
    expect(ast.acceptance_criteria).toHaveLength(1);
    expect(ast.acceptance_criteria[0].id).toBe('AC-01');
  });

  // @ac AC-02: Spec missing required field 'id' returns ParseError with field path 'spec.id'
  // References: C-01 (schema validation), C-02 (error with field path)
  it('AC-02: returns error with path "spec.id" when id is missing', () => {
    const yaml = readFixture('invalid/missing-id.spec.yaml');
    const result = parseSpec(yaml);

    expect(result.ok).toBe(false);
    if (result.ok) return;

    const idError = result.errors.find((e) => e.path.includes('id'));
    expect(idError).toBeDefined();
    expect(idError!.path).toBe('spec.id');
    expect(idError!.type).toBe('required');
  });

  // @ac AC-03: Spec with unknown field returns ParseError identifying the extra field
  // References: C-03 (additionalProperties enforcement)
  it('AC-03: rejects unknown fields with additionalProperties error', () => {
    const yaml = readFixture('invalid/extra-field.spec.yaml');
    const result = parseSpec(yaml);

    expect(result.ok).toBe(false);
    if (result.ok) return;

    const extraError = result.errors.find((e) => e.type === 'additionalProperties');
    expect(extraError).toBeDefined();
    expect(extraError!.path).toContain('unknown_field');
  });

  // @ac AC-04: Malformed YAML (bad indentation) returns ParseError with line number
  // References: C-02 (line numbers), C-05 (graceful error handling)
  it('AC-04: handles malformed YAML gracefully with line number', () => {
    const yaml = readFixture('invalid/bad-yaml.spec.yaml');
    const result = parseSpec(yaml);

    expect(result.ok).toBe(false);
    if (result.ok) return;

    expect(result.errors.length).toBeGreaterThan(0);
    expect(result.errors[0].type).toBe('yaml_syntax');
    expect(result.errors[0].line).toBeDefined();
    expect(typeof result.errors[0].line).toBe('number');
  });

  // @ac AC-05: Spec with invalid version format returns error
  // References: C-01 (schema validation)
  it('AC-05: rejects invalid version format with pattern error', () => {
    const yaml = readFixture('invalid/bad-version.spec.yaml');
    const result = parseSpec(yaml);

    expect(result.ok).toBe(false);
    if (result.ok) return;

    const versionError = result.errors.find((e) => e.path === 'spec.version');
    expect(versionError).toBeDefined();
    expect(versionError!.type).toBe('pattern');
  });

  // @ac AC-06: Spec with all optional fields omitted parses successfully
  // References: C-01 (schema validation), C-04 (typed SpecAST)
  it('AC-06: parses minimal spec with only required fields', () => {
    const yaml = readFixture('valid/minimal.spec.yaml');
    const result = parseSpec(yaml);

    expect(result.ok).toBe(true);
    if (!result.ok) return;

    const ast = result.value;
    expect(ast.id).toBe('test-minimal');
    expect(ast.depends_on).toBeUndefined();
    expect(ast.trust_level).toBeUndefined();
    expect(ast.environment).toBeUndefined();
    expect(ast.tags).toBeUndefined();
    expect(ast.changelog).toBeUndefined();
    expect(ast.generated_from).toBeUndefined();
  });

  // @ac AC-07: Spec using YAML anchors and aliases is parsed with anchors resolved
  // References: C-06 (YAML anchors/aliases)
  it('AC-07: resolves YAML anchors and aliases correctly', () => {
    const yaml = readFixture('valid/with-anchors.spec.yaml');
    const result = parseSpec(yaml);

    expect(result.ok).toBe(true);
    if (!result.ok) return;

    const ast = result.value;
    // Both constraints should have type and enforcement from the anchor
    expect(ast.constraints[0].type).toBe('technical');
    expect(ast.constraints[0].enforcement).toBe('error');
    expect(ast.constraints[1].type).toBe('technical');
    expect(ast.constraints[1].enforcement).toBe('error');
  });

  // @ac AC-08: Spec with multiple errors returns all errors, not just the first
  // References: C-02 (error reporting), C-07 (collect all errors)
  it('AC-08: collects multiple validation errors', () => {
    const yaml = readFixture('invalid/multiple-errors.spec.yaml');
    const result = parseSpec(yaml);

    expect(result.ok).toBe(false);
    if (result.ok) return;

    // Should have at least 2 errors: missing id + invalid version pattern
    expect(result.errors.length).toBeGreaterThanOrEqual(2);
  });

  // @ac AC-09: Invalid constraint ID format returns error
  // References: C-01 (schema validation)
  it('AC-09: rejects invalid constraint ID format', () => {
    const yaml = readFixture('invalid/bad-constraint-id.spec.yaml');
    const result = parseSpec(yaml);

    expect(result.ok).toBe(false);
    if (result.ok) return;

    const constraintError = result.errors.find(
      (e) => e.path.includes('constraints') && e.type === 'pattern',
    );
    expect(constraintError).toBeDefined();
  });

  // @ac AC-10: Invalid AC ID format returns error
  // References: C-01 (schema validation)
  it('AC-10: rejects invalid AC ID format', () => {
    const yaml = readFixture('invalid/bad-ac-id.spec.yaml');
    const result = parseSpec(yaml);

    expect(result.ok).toBe(false);
    if (result.ok) return;

    const acError = result.errors.find(
      (e) => e.path.includes('acceptance_criteria') && e.type === 'pattern',
    );
    expect(acError).toBeDefined();
  });

  // Additional structural test: C-08 — pure function, no side effects
  it('C-08: parseSpec is a pure function (no I/O, no side effects)', () => {
    // Calling with the same input should produce the same output
    const yaml = readFixture('valid/simple.spec.yaml');
    const result1 = parseSpec(yaml);
    const result2 = parseSpec(yaml);

    expect(result1).toEqual(result2);
  });
});

File: specter/internal/parser/parse_test.go
// @spec spec-parse
package parser

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func readFixture(t *testing.T, relPath string) string {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "..", "testdata", relPath))
	if err != nil {
		t.Fatalf("failed to read fixture %s: %v", relPath, err)
	}
	return string(data)
}

// @ac AC-01
func TestParseValidSpec(t *testing.T) {
	yaml := readFixture(t, "valid/simple.spec.yaml")
	result := ParseSpec(yaml)

	if !result.OK {
		t.Fatalf("expected OK, got errors: %v", result.Errors)
	}
	if result.Value.ID != "test-simple" {
		t.Errorf("expected id 'test-simple', got %q", result.Value.ID)
	}
	if result.Value.Version != "1.0.0" {
		t.Errorf("expected version '1.0.0', got %q", result.Value.Version)
	}
	if result.Value.Status != "approved" {
		t.Errorf("expected status 'approved', got %q", result.Value.Status)
	}
	if result.Value.Tier != 2 {
		t.Errorf("expected tier 2, got %d", result.Value.Tier)
	}
	if result.Value.Context.System != "Test system" {
		t.Errorf("expected system 'Test system', got %q", result.Value.Context.System)
	}
	if len(result.Value.Constraints) != 1 {
		t.Errorf("expected 1 constraint, got %d", len(result.Value.Constraints))
	}
	if len(result.Value.AcceptanceCriteria) != 1 {
		t.Errorf("expected 1 AC, got %d", len(result.Value.AcceptanceCriteria))
	}
}

// @ac AC-02
func TestParseMissingID(t *testing.T) {
	yaml := readFixture(t, "invalid/missing-id.spec.yaml")
	result := ParseSpec(yaml)

	if result.OK {
		t.Fatal("expected failure, got OK")
	}
	found := false
	for _, e := range result.Errors {
		if e.Type == "required" {
			found = true
		}
	}
	if !found {
		t.Errorf("expected 'required' error type, got: %v", result.Errors)
	}
}

// @ac AC-03
func TestParseExtraField(t *testing.T) {
	yaml := readFixture(t, "invalid/extra-field.spec.yaml")
	result := ParseSpec(yaml)

	if result.OK {
		t.Fatal("expected failure, got OK")
	}
	found := false
	for _, e := range result.Errors {
		if e.Type == "additionalProperties" {
			found = true
		}
	}
	if !found {
		t.Errorf("expected 'additionalProperties' error, got: %v", result.Errors)
	}
}

// @ac AC-04
func TestParseBadYAML(t *testing.T) {
	yaml := readFixture(t, "invalid/bad-yaml.spec.yaml")
	result := ParseSpec(yaml)

	if result.OK {
		t.Fatal("expected failure, got OK")
	}
	if len(result.Errors) == 0 {
		t.Fatal("expected errors")
	}
}

// @ac AC-05
func TestParseBadVersion(t *testing.T) {
	yaml := readFixture(t, "invalid/bad-version.spec.yaml")
	result := ParseSpec(yaml)

	if result.OK {
		t.Fatal("expected failure, got OK")
	}
	found := false
	for _, e := range result.Errors {
		if e.Type == "pattern" {
			found = true
		}
	}
	if !found {
		t.Errorf("expected 'pattern' error, got: %v", result.Errors)
	}
}

// @ac AC-06
func TestParseMinimalSpec(t *testing.T) {
	yaml := readFixture(t, "valid/minimal.spec.yaml")
	result := ParseSpec(yaml)

	if !result.OK {
		t.Fatalf("expected OK, got errors: %v", result.Errors)
	}
	if result.Value.ID != "test-minimal" {
		t.Errorf("expected id 'test-minimal', got %q", result.Value.ID)
	}
	if result.Value.DependsOn != nil {
		t.Error("expected nil depends_on")
	}
	if result.Value.Tags != nil {
		t.Error("expected nil tags")
	}
}

// @ac AC-07
func TestParseWithAnchors(t *testing.T) {
	yaml := readFixture(t, "valid/with-anchors.spec.yaml")
	result := ParseSpec(yaml)

	if !result.OK {
		t.Fatalf("expected OK, got errors: %v", result.Errors)
	}
	if len(result.Value.Constraints) < 2 {
		t.Fatalf("expected at least 2 constraints, got %d", len(result.Value.Constraints))
	}
	if result.Value.Constraints[0].Type != "technical" {
		t.Errorf("expected constraint type 'technical', got %q", result.Value.Constraints[0].Type)
	}
	if result.Value.Constraints[1].Type != "technical" {
		t.Errorf("expected constraint type 'technical', got %q", result.Value.Constraints[1].Type)
	}
}

// @ac AC-08
func TestParseMultipleErrors(t *testing.T) {
	yaml := readFixture(t, "invalid/multiple-errors.spec.yaml")
	result := ParseSpec(yaml)

	if result.OK {
		t.Fatal("expected failure, got OK")
	}
	if len(result.Errors) < 2 {
		t.Errorf("expected at least 2 errors, got %d: %v", len(result.Errors), result.Errors)
	}
}

// @ac AC-09
func TestParseBadConstraintID(t *testing.T) {
	yaml := readFixture(t, "invalid/bad-constraint-id.spec.yaml")
	result := ParseSpec(yaml)

	if result.OK {
		t.Fatal("expected failure, got OK")
	}
	found := false
	for _, e := range result.Errors {
		if e.Type == "pattern" {
			found = true
		}
	}
	if !found {
		t.Errorf("expected 'pattern' error for constraint ID, got: %v", result.Errors)
	}
}

// @ac AC-10
func TestParseBadACID(t *testing.T) {
	yaml := readFixture(t, "invalid/bad-ac-id.spec.yaml")
	result := ParseSpec(yaml)

	if result.OK {
		t.Fatal("expected failure, got OK")
	}
	found := false
	for _, e := range result.Errors {
		if e.Type == "pattern" {
			found = true
		}
	}
	if !found {
		t.Errorf("expected 'pattern' error for AC ID, got: %v", result.Errors)
	}
}

// @ac AC-11
func TestParseHumanReadable_ConstraintID(t *testing.T) {
	yaml := readFixture(t, "invalid/bad-constraint-id.spec.yaml")
	result := ParseSpec(yaml)

	if result.OK {
		t.Fatal("expected failure, got OK")
	}
	for _, e := range result.Errors {
		if e.Type == "pattern" {
			if !strings.Contains(e.Message, "C-NN") {
				t.Errorf("expected message to mention C-NN pattern, got: %q", e.Message)
			}
			return
		}
	}
	t.Errorf("no pattern error found: %v", result.Errors)
}

// @ac AC-12
func TestParseHumanReadable_MissingID(t *testing.T) {
	yaml := readFixture(t, "invalid/missing-id.spec.yaml")
	result := ParseSpec(yaml)

	if result.OK {
		t.Fatal("expected failure, got OK")
	}
	for _, e := range result.Errors {
		if e.Type == "required" {
			if !strings.Contains(e.Message, "kebab-case") {
				t.Errorf("expected message to mention kebab-case, got: %q", e.Message)
			}
			return
		}
	}
	t.Errorf("no required error found: %v", result.Errors)
}

// @ac AC-13
func TestParseHumanReadable_ExtraField(t *testing.T) {
	yaml := readFixture(t, "invalid/extra-field.spec.yaml")
	result := ParseSpec(yaml)

	if result.OK {
		t.Fatal("expected failure, got OK")
	}
	for _, e := range result.Errors {
		if e.Type == "additionalProperties" {
			if strings.Contains(e.Message, "additionalProperties") {
				t.Errorf("message should not expose raw 'additionalProperties', got: %q", e.Message)
			}
			return
		}
	}
	t.Errorf("no additionalProperties error found: %v", result.Errors)
}

func TestParsePureFunction(t *testing.T) {
	yaml := readFixture(t, "valid/simple.spec.yaml")
	r1 := ParseSpec(yaml)
	r2 := ParseSpec(yaml)

	if r1.OK != r2.OK {
		t.Error("pure function violation: different OK values")
	}
	if r1.Value.ID != r2.Value.ID {
		t.Error("pure function violation: different IDs")
	}
}

// @ac AC-14 (v0.7.0 — context.additionalProperties tightened to false)
func TestParse_UnknownContextField_Rejected(t *testing.T) {
	yaml := `spec:
  id: test-unknown-context
  version: "1.0.0"
  status: draft
  tier: 3
  context:
    system: test
    role: "this field is not in the schema"
  objective:
    summary: test
  constraints:
    - id: C-01
      description: "test"
  acceptance_criteria:
    - id: AC-01
      description: "test"
      references_constraints: ["C-01"]
`
	result := ParseSpec(yaml)
	if result.OK {
		t.Fatal("expected parse to fail on unknown context field")
	}
	found := false
	for _, e := range result.Errors {
		if strings.Contains(e.Message, "role") || strings.Contains(e.Path, "context") {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("expected error mentioning 'role' or 'context', got: %v", result.Errors)
	}
}

// @ac AC-15 (v0.7.0 — AC metadata fields)
func TestParse_ACNotesAndApprovalFields(t *testing.T) {
	yaml := `spec:
  id: test-ac-metadata
  version: "1.0.0"
  status: approved
  tier: 1
  context:
    system: test
  objective:
    summary: test
  constraints:
    - id: C-01
      description: "test constraint"
  acceptance_criteria:
    - id: AC-01
      description: "test AC"
      references_constraints: ["C-01"]
      notes: "Financial op — see also AC-03."
      approval_gate: true
      approval_date: "2026-04-17"
`
	result := ParseSpec(yaml)
	if !result.OK {
		t.Fatalf("expected OK, got errors: %v", result.Errors)
	}
	ac := result.Value.AcceptanceCriteria[0]
	if ac.Notes != "Financial op — see also AC-03." {
		t.Errorf("Notes not preserved, got %q", ac.Notes)
	}
	if !ac.ApprovalGate {
		t.Error("ApprovalGate should be true")
	}
	if ac.ApprovalDate != "2026-04-17" {
		t.Errorf("ApprovalDate mismatch, got %q", ac.ApprovalDate)
	}
}

// @ac AC-15
func TestParse_ACApprovalDateInvalidFormat_Rejected(t *testing.T) {
	yaml := `spec:
  id: test-bad-date
  version: "1.0.0"
  status: approved
  tier: 1
  context:
    system: test
  objective:
    summary: test
  constraints:
    - id: C-01
      description: "test"
  acceptance_criteria:
    - id: AC-01
      description: "test"
      references_constraints: ["C-01"]
      approval_gate: true
      approval_date: "not-a-date"
`
	result := ParseSpec(yaml)
	if result.OK {
		t.Fatal("expected parse to fail on invalid approval_date format")
	}
}

// @ac AC-16 (v0.7.0 — parse-time cross-reference validation)
func TestParse_DanglingConstraintReference_Rejected(t *testing.T) {
	yaml := `spec:
  id: test-dangling
  version: "1.0.0"
  status: draft
  tier: 3
  context:
    system: test
  objective:
    summary: test
  constraints:
    - id: C-01
      description: "only declared constraint"
  acceptance_criteria:
    - id: AC-01
      description: "references something real"
      references_constraints: ["C-01"]
    - id: AC-02
      description: "references something fake"
      references_constraints: ["C-99"]
`
	result := ParseSpec(yaml)
	if result.OK {
		t.Fatal("expected parse to fail on dangling reference")
	}
	found := false
	for _, e := range result.Errors {
		if e.Type == "dangling_reference" && strings.Contains(e.Message, "C-99") {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("expected dangling_reference error mentioning C-99, got: %v", result.Errors)
	}
}
